rusk-wallet/src/tx.rs

```diff
@@ -0,0 +1,729 @@
+use dusk_plonk::prelude::{JubJubScalar, Proof};
+use dusk_poseidon::cipher::PoseidonCipher;
+use dusk_poseidon::sponge::hash;
+use dusk_poseidon::tree::PoseidonBranch;
+use dusk_schnorr::Proof as SchnorrSig;
+use phoenix_core::{Crossover, Fee, Note};
+use rand_core::{CryptoRng, RngCore};
+
+const CONTRACT_ID_SIZE: usize = mem::size_of::<ContractId>();
+
+/// The structure sent over the network representing a transaction.
+#[derive(Debug, Clone)]
+pub struct Transaction {
+    nullifiers: Vec<BlsScalar>,
+    outputs: Vec<Note>,
+    anchor: BlsScalar,
+    proof: Proof,
+    fee: Fee,
+    crossover: Crossover,
+    call: Option<(ContractId, Vec<u8>)>,
+}
+
+impl Transaction {
+    /// Creates a transaction from the skeleton and the proof.
+    fn new(tx_skel: TransactionSkeleton, proof: Proof) -> Self {
+        Self {
+            proof,
+            nullifiers: tx_skel.nullifiers,
+            outputs: tx_skel.outputs,
+            anchor: tx_skel.anchor,
+            fee: tx_skel.fee,
+            crossover: tx_skel.crossover,
+            call: tx_skel.call,
+        }
+    }
+
+    /// Hashes the transaction excluding.
+    pub fn hash(&self) -> BlsScalar {
+        let skel = TransactionSkeleton::from(self.clone());
+        skel.hash()
+    }
+
+    /// Returns the inputs to the hash function.
+    pub fn hash_inputs(&self) -> Vec<BlsScalar> {
+        let skel = TransactionSkeleton::from(self.clone());
+        skel.hash_inputs()
+    }
+
+    /// Serializes the transaction into a variable length byte buffer.
+    pub fn to_var_bytes(&self) -> Result<Vec<u8>, BytesError> {
+        // compute the serialized size to preallocate space
+        let size = u64::SIZE
+            + self.nullifiers.len() * BlsScalar::SIZE
+            + u64::SIZE
+            + self.outputs.len() * Note::SIZE
+            + BlsScalar::SIZE
+            + Fee::SIZE
+            + Proof::SIZE
+            + Crossover::SIZE
+            + u64::SIZE
+            + self
+                .call
+                .as_ref()
+                .map(|(_, cdata)| CONTRACT_ID_SIZE + cdata.len())
+                .unwrap_or(0);
+
+        let mut bytes = vec![0u8; size];
+        let mut writer = &mut bytes[..];
+
+        writer.write(&(self.nullifiers.len() as u64).to_bytes())?;
+        for input in &self.nullifiers {
+            writer.write(&input.to_bytes())?;
+        }
+
+        writer.write(&(self.outputs.len() as u64).to_bytes())?;
+        for output in &self.outputs {
+            writer.write(&output.to_bytes())?;
+        }
+
+        writer.write(&self.anchor.to_bytes())?;
+        writer.write(&self.fee.to_bytes())?;
+        writer.write(&self.proof.to_bytes())?;
+        writer.write(&self.crossover.to_bytes())?;
+
+        write_optional_call(&mut writer, self.call.as_ref())?;
+
+        Ok(bytes)
+    }
+
+    /// Deserializes the transaction from a bytes buffer.
+    pub fn from_bytes<B: AsRef<[u8]>>(buf: B) -> Result<Self, BytesError> {
```

Comment:
I did this one just out of habit honestly. Since the entire stack seems to do it the `&[u8]` way, I'll do that. In general it's nice to be able to pass fixed size buffers as well.